utils/pattern/src/frontend/serde.rs

```diff
@@ -5,70 +5,163 @@
 use super::*;
 use alloc::borrow::Cow;
 use alloc::vec::Vec;
-use core::fmt::Display;
 
 use ::serde::{Deserialize, Deserializer, Serialize, Serializer};
 
+#[doc(hidden)]
+pub fn deserialize_option_borrowed_cow<'de, D, B>(
+    deserializer: D,
+) -> Result<Option<Cow<'de, Pattern<B>>>, D::Error>
+where
+    D: Deserializer<'de>,
+    B: PatternBackend<Store = str>,
+    B::PlaceholderKeyCow<'de>: Deserialize<'de>,
+{
+    #[derive(Deserialize)]
+    #[serde(transparent, bound = "'data: 'de")]
+    // Cows fail to borrow in some situations (array, option), but structs of Cows don't.
+    struct CowPatternWrap<'data, B: PatternBackend<Store = str>>(
+        #[serde(borrow, deserialize_with = "deserialize_borrowed_cow")] pub Cow<'data, Pattern<B>>,
+    )
+    where
+        Pattern<B>: ToOwned,
+        B: PatternBackend<Store = str>,
+        B::PlaceholderKeyCow<'data>: Deserialize<'data>;
+
+    Option::<CowPatternWrap<'de, B>>::deserialize(deserializer)
+        .map(|array| array.map(|wrap| wrap.0))
+}
+
 type HumanReadablePattern<'a, B> =
     Vec<PatternItemCow<'a, <B as PatternBackend>::PlaceholderKeyCow<'a>>>;
 
-impl<'de, 'data, B, Store, E> Deserialize<'de> for Pattern<B, Store>
+#[derive(Debug, PartialEq)]
+#[allow(clippy::exhaustive_structs)] // newtype
+pub struct PatternString<P: PatternBackend>(pub Box<Pattern<P>>);
+
+impl<B: PatternBackend> Clone for PatternString<B>
+where
+    Box<B::Store>: for<'a> From<&'a B::Store>,
+{
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl<P: PatternBackend> core::ops::Deref for PatternString<P> {
+    type Target = Pattern<P>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<B: PatternBackend> Default for PatternString<B>
+where
+    B: PatternBackend,
+    for<'a> B::PlaceholderKeyCow<'a>: core::str::FromStr,
+    for<'a> <B::PlaceholderKeyCow<'a> as core::str::FromStr>::Err: core::fmt::Debug,
+{
+    fn default() -> Self {
+        #[allow(clippy::unwrap_used)] // todo does this work?
+        Self(Pattern::<B>::try_from_str("", Default::default()).unwrap())
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de, B: PatternBackend> serde::Deserialize<'de> for PatternString<B>
+where
+    B: PatternBackend,
+    B::PlaceholderKeyCow<'de>: core::str::FromStr,
+    <B::PlaceholderKeyCow<'de> as core::str::FromStr>::Err: core::fmt::Debug,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        Ok(Self(
+            Pattern::<B>::try_from_str(&<Cow<str>>::deserialize(deserializer)?, Default::default())
```

Comment:
Nit: I think `<Cow<str>>::deserialize` always deserializes to a String. `serde_derive` has special-case code that essentially adds a `deserialize_with` if Syn sees a `Cow<str>`. It doesn't actually specialize `Cow<str>` to borrow when possible.